relay/shfu.go

```diff
@@ -0,0 +1,149 @@
+	if err != nil {
+		return nil, fmt.Errorf("failed to get latest finalized header: %w", err)
+	}
+
+	toHeight := latestFinalizedHeader.GetHeight()
+
+	// Check if we already have a record with toHeight >= current toHeight
+	latestRecord, err := storage.GetLatestSHFUForChain(ctx, target.ChainID(), counterparty.ChainID())
+	if err != nil {
+		return nil, fmt.Errorf("failed to get latest SHFU record: %w", err)
+	}
+	if latestRecord != nil && !latestRecord.ToHeight.LT(toHeight) {
+		// We already have a record with toHeight >= current toHeight, so skip execution
+		logger.InfoContext(ctx, "Skipping SHFU execution: already have record with sufficient height",
+			"chain_id", target.ChainID(),
+			"counterparty_chain_id", counterparty.ChainID(),
+			"current_to_height", toHeight.String(),
+			"existing_to_height", latestRecord.ToHeight.String())
+		return nil, nil
+	}
+
+	// Call setupHeadersForUpdate0 with the counterparty chain
+	results, err := lcpProver.setupHeadersForUpdate0(ctx, counterparty, latestFinalizedHeader)
+	if err != nil {
+		return nil, fmt.Errorf("failed to call setupHeadersForUpdate0: %w", err)
+	}
+
+	h2h := func(h ibcexported.Height) clienttypes.Height {
+		return clienttypes.Height{
+			RevisionNumber: h.GetRevisionNumber(),
+			RevisionHeight: h.GetRevisionHeight(),
+		}
+	}
+
+	// Serialize the latestFinalizedHeader (ClientMessage) to bytes
+	anyMsg, err := types.NewAnyWithValue(latestFinalizedHeader)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create Any from latestFinalizedHeader: %w", err)
+	}
+
+	latestFinalizedHeaderBytes, err := target.Codec().Marshal(anyMsg)
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal latestFinalizedHeader: %w", err)
+	}
+
+	// Create SHFU record for database storage
+	record := &SHFURecord{
+		ChainID:               target.ChainID(),
+		CounterpartyChainID:   counterparty.ChainID(),
+		ToHeight:              h2h(toHeight),
+		ToHeightTime:          time.Now(), // Could be extracted from header if available
+		UpdatedAt:             time.Now(),
+		UpdateClientResults:   results,
+		LatestFinalizedHeader: latestFinalizedHeaderBytes,
+	}
+
+	logger.InfoContext(ctx, "SHFU executed successfully",
+		"target_chain_id", target.ChainID(),
+		"counterparty_chain_id", counterparty.ChainID())
+
+	// Save the record to database with retry logic for temporary errors
+	err = retry.Do(
+		func() error {
+			return storage.SaveSHFUResult(ctx, record)
+		},
+		retry.Context(ctx),
+		retry.Attempts(3),
+		retry.Delay(100*time.Millisecond),
+		retry.DelayType(retry.BackOffDelay),
+		retry.RetryIf(func(err error) bool {
+			// Retry only if it's a temporary error according to the storage implementation
+			return storage.IsTemporaryError(err)
+		}),
+		retry.OnRetry(func(n uint, err error) {
+			logger.ErrorContext(ctx, "Retry attempt for SaveSHFUResult due to temporary error", err,
+				"attempt", n+1,
+				"target_chain_id", target.ChainID())
```

Comment:
`counterparty_chain_id` information would be helpful:

```suggestion
				"target_chain_id", target.ChainID(),
				"counterparty_chain_id", counterparty.ChainID())				
```

Reply:
fixed